### DISCUSSION TITLE
Compilation

### DISCUSSION THREAD

**Discussion by ghost** [evilsocket/opensnitch#933] [General]
Hi.
Trying to follow this instruction: https://github.com/evilsocket/opensnitch/wiki/Compilation

```
sudo apt-get install git golang libnetfilter-queue-dev libpcap-dev protobuf-compiler python3-pip pyqt5-dev-tools qttools5-dev-tools qtbase5-dev qtchooser qt5-qmake qtbase5-dev-tools python-pyqt5.qtsql python3-notify2
```

1. We dont need to `install golang`, we have a note `Make sure you have a correctly configured Go >= 1.15 environment`. Installing go this way may brake existing go setup.
2. Note suggests `Go >= 1.15` but `daemon/go.mod` says `go 1.14`
3. There package missing in debian testing (bookworm): `python-pyqt5.qtsql`, it exists as `python3-pyqt5.qtsql`

What I suggest:
```
sudo apt install git libnetfilter-queue-dev libpcap-dev protobuf-compiler python3-pip pyqt5-dev-tools qttools5-dev-tools qtbase5-dev qtchooser qt5-qmake qtbase5-dev-tools python3-pyqt5.qtsql python3-notify2
```

Next:
```
python3 -m pip install --user grpcio-tools qt-material
```
On fresh debian install gives me:
```
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.
    
    If you wish to install a non-Debian packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.
    
    See /usr/share/doc/python3.11/README.venv for more information.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
```
I installed by venv way, error is gone, but this error will brake `make` later.

After that:

```
# make
go: github.com/evilsocket/ftrace@v1.2.0: missing go.sum entry; to add it:
	go mod download github.com/evilsocket/ftrace
go: github.com/evilsocket/ftrace@v1.2.0: missing go.sum entry; to add it:
	go mod download github.com/evilsocket/ftrace
make: *** [Makefile:12: opensnitchd] Error 1

```
There is no `go.sum` in `daemon` at all, so we will get errors for all dependencies.
After `go mod tidy` in `daemon` there is two dependencies removed:
```
github.com/vishvananda/netns v0.0.0-20191106174202-0a2b9b5464df // indirect
golang.org/x/sys v0.0.0-20190606203320-7fc4e5ec1444 // indirect
```
And my IDE suggest that two dependencies is vurnerable:
```
golang.org/x/net v0.0.0-20180417003750-8d16fa6dc9a8
golang.org/x/text v0.3.0 // indirect
```
I think it will be a good idea to update dependencies.
After that:
```
make[2]: Entering directory '/home/user/code/ring-c/opensnitch/ui/i18n'
/usr/lib/python3/dist-packages/PyQt5/pylupdate_main.py:73: DeprecationWarning: Use setlocale(), getencoding() and getlocale() instead
  encoding = locale.getdefaultlocale()[1]
```
And I get a `error: externally-managed-environment` again.
I can make a PR for go and wiki stuff, but what to do with python errors?

**Comment by gustavo-iniguez-goya** [comment]
Hi @Ring-C ,

The compilation instructions are not tied to a specific distribution. As stated below the title: `(tested on Debian Sid and Ubuntu 18 and 20. It may fail on other distributions.)`

By compiling the project I'd assume that you know what you're doing.

I don't understand the point 1), what do you mean by "We dont need to install golang" ?
2) we should even be able to compile with Go 1.13 as far as I can remember.
Point 3) fixed, my fault, sorry O:)

> python3 -m pip install --user grpcio-tools qt-material

That step is optional on Debian >= Bullseye. You can install python3-grpcio and python3-protobuf. However some distributions didn't package those libs, so that's why it was added to the wiki.
Nowadays I think most distros I've tried package them already, so we can move the pip3 command to a note.

Regarding the warning, it's a policy change that have been recently added: https://www.omgubuntu.co.uk/2023/04/pip-install-error-externally-managed-environment-fix 

> go mod download github.com/evilsocket/ftrace

We removed that dependency some months ago. We don't use it anymore. On Debian Sid it compiles just fine :)
